### ISSUE TITLE
Need better words to conceptually describe states of BindingsSets

### ISSUE THREAD

**Issue by luketpeterson** [trueagi-io/hyperon-experimental#281]
From @vsbogd  (in code review)
> BindingsSet::single in matcher means that two atoms are equal, and BindingsSet::empty means they cannot be matched. Thus I don't see the use-case for such semantics of _empty() yet. And if this semantics is not used then name can be confusing because one may think it returns true iff no Bindings instances are inside.

I don't feel like this conceptual difference is adequately communicated by the words "empty" and "single".

Maybe "unconstrained" and "fully_constrained"?  Those are clumsy to write and probably not accurate.  Any other ideas?

**Comment by vsbogd** [comment]
> Maybe "unconstrained" and "fully_constrained"? Those are clumsy to write and probably not accurate. Any other ideas?

The instance of `Bindings` is a container for a single set of variable values and equalities. When we match (maybe unify is a better term here, but match is used in the code) two atoms, we can receive a number of such containers (instances of `Bindings`) as a result. Each instance of `Bindings` in this set can be considered as independent from others.

There are three kinds of results:
1. If matching is not successful, then the result is empty set of `Bindings`. 
2. If two atoms being matched are equal, then the result contains a single instance of `Bindings` which doesn't have variables inside.
3. If there are many different options to match two atoms, then the result contains a set of `Bindings`, each instance of `Bindings` contains the unique set of variables and values.

The result of matching can also be represented by a generator instead of a finite set, and it is one of the reasons the `match_atoms` returns the instance of `Iterator`. For now the `BindingsSet` represents the result as a finite set of `Bindings`.

I would like to make the points here:
- The result of the atom matching is not a single set of variable values (i.e. `Bindings`), but a set of such sets;
- User need the ability to construct empty instance of the `BindingsSet` to return it when matching is unsuccessful (situation (1)), it is needed to implement a custom matching interface;
- I am not sure `unconstrained` and `fully_constrained` reflect the situations (1) and (2) respectively.
 
Current names are bound to the set nature of the `BindingsSet`. Name `empty()` reads more or less ok, but it is not clear for the developer who is out of context. It constructs the new "empty" instance of the `BindingsSet`, so in Rust conventions it should be `BindingsSet::new()` instead. Then empty instance of `BindingsSet` can be returned as a result for the situation (1) or extended to the situation (2) or (3). `single()` is definitely unclear because single element of the set could be anything. And here we specifically use this constructor to represent the situation (2). May be we don't need `single()` at all and should use chain of calls something like `BindingsSet::new().add_full_match()`.

Luke, could you please explain the logic you have under the names you suggested?

**Comment by vsbogd** [comment]
> I'm seeing this as a teeny tiny step towards laying the foundation to support infinite, explosive, lazily evaluated, or otherwise unbounded bindings.

I agree that it is would be more general to implement `BindingsSet` as a stream or comprehension of the `Bindings` instances. We also can replace the `MatchResultIter` by `BindingsSet`. But to make it work we need to replace most of the `Vec` in the implementation by the streams or comprehensions. It is not easy on the current implementation but can be more easy after moving on the minimal interpreter implementation I am trying to evaluate at the moment.

**Comment by luketpeterson** [comment]
I have been hesitant to reply here because I feel that what I need is a good mental model for what a Bindings (and also a BindingsSet) is at a conceptual level.  I understand what it means within the context of the current MeTTa hyperon-experimental implementation, but that is only a specific instance of a larger concept.  For example, I think of a Bindings as a container for *Information* in the sense of `Information Programming` proposed by https://github.com/Adam-Vandervorst .

Anyway, there is more to say on this topic, but I wanted to mention this issue in the context of https://github.com/trueagi-io/hyperon-experimental/pull/298.  If we can establish a "North Star" about what form Bindings take at a conceptual level (possibly an amalgamated form that allows for specific bindings a'la the current Bindings type, in addition to overarching rules) then I think this will help answer the questions down in the weeds about what a Bindings API should look like.  Sorry I don't have a specific proposal yet.  Still learning the solution space.

**Comment by vsbogd** [comment]
Am I right that your idea is to replace `BindingsSet` by something more abstract like "Information"? In such a case think we need to talk about interpreter and its principles rather then talk about `Bindings` or `BindingsSet`. `Bindings` has a very specific role in a context of the current interpreter: keeping values of variables and relations between variables. But it is not the complete information interpreter operates.  It also has the current result of the atom transformations for instance.

**Comment by luketpeterson** [comment]
My thoughts aren't formed to the point yet that I could propose specific changes RE architecture.  So I wouldn't say I have an idea.  But yes, it seems to me that Bindings / BindingsSet is a container for intermediate information representing the results of a space query operation or a grounded atom match operation.  So in theory that container could be reworked to contain more abstract information.  Unfortunately I don't have any concrete proposals yet.
